### ISSUE TITLE
Upgrade from 4.14 to 4.15 fails when using XenServer 7.0/7.1/7.4

### ISSUE THREAD

**Issue by vladimirpetrov** [apache/cloudstack#4502]
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
~~~
Upgrade
~~~

##### CLOUDSTACK VERSION
~~~
Upgrade from 4.14 to 4.15
~~~

##### CONFIGURATION
2 hosts using XenServer 7.0/7.1/7.4

##### OS / ENVIRONMENT
XenServer 7.0(7.1, 7.4), two hypervisor hosts, CentOS 7.6 as management server OS.

##### SUMMARY
After upgrading from 4.14 to 4.15 and destroying the system VMs, they are not able to start again.

##### STEPS TO REPRODUCE
~~~
1. Install CloudStack 4.14 with XenServer 7.0 (7.1, 7.4), create some test objects - VMs, snapshots, networks, etc.
2. Follow the standard upgrade procedure
3. After the upgrade, destroy the system VMs (as described in the documentation)
~~~

##### EXPECTED RESULTS

~~~
Successful upgrade.
~~~

##### ACTUAL RESULTS
~~~
The new system VMs are not able to start, here is what is shown in the log:

2020-11-18 21:48:00,312 ERROR [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-8:ctx-5c32b189 job-71/job-72 ctx-c90dd860) (logid:84ab8c0f) Failed to start instance VM[DomainRouter|r-14-VM]
com.cloud.utils.exception.ExecutionException: Unable to start  VM:4acfa807-b78e-48f0-b615-177c9a9c954d due to error in finalizeStart, not retrying
        at com.cloud.vm.VirtualMachineManagerImpl.orchestrateStart(VirtualMachineManagerImpl.java:1266)
        at com.cloud.vm.VirtualMachineManagerImpl.orchestrateStart(VirtualMachineManagerImpl.java:5461)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.cloud.vm.VmWorkJobHandlerProxy.handleVmWorkJob(VmWorkJobHandlerProxy.java:107)
        at com.cloud.vm.VirtualMachineManagerImpl.handleVmWorkJob(VirtualMachineManagerImpl.java:5628)
        at com.cloud.vm.VmWorkJobDispatcher.runJob(VmWorkJobDispatcher.java:102)
        at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:620)
        at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:48)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
        at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:45)
        at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:568)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)

After some investigations with @rhtyd , it turned out the reason is that the systemvm.iso file is not detected by the host VM.
~~~

**Comment by vladimirpetrov** [comment]
I just tested that the issue is reproducible also with XenServer 7.4.

**Comment by rhtyd** [comment]
@vladimirpetrov to confirm regression can you do an upgrade of a XenServer 7.1 env from 4.11 or 4.13 to 4.14 and see how the system behaves? If you get the same issue we can lower the severity and address workaround in upgrade docs.

**Comment by vladimirpetrov** [comment]
ACS 4.11 -> 4.14
ACS 4.13 -> 4.14
ACS 4.11 -> 4.15 and
ACS 4.13 -> 4.15 (all with XenServer 7.1) upgrades passed successfully.

**Comment by andrijapanicsb** [comment]
Can you share more logs/info to let others also understand the issue as well.

Also, I see " _CentOS 7.0 on the hosts._" in the description, but you are saying it's the XS setup - do you have mixed hypervisors or why is there " CentOS 7.0 on the hosts." ?

**Comment by DaanHoogland** [comment]
> ACS 4.11 -> 4.14
> ACS 4.13 -> 4.14
> ACS 4.11 -> 4.15 and
> ACS 4.13 -> 4.15 (all with XenServer 7.1) upgrades passed successfully.

given that all these succeed and 4.14 -> 4.15 fail, there is likely to be something wrong in the upgrade path. Can you share the log of the initial start after upgrade @vladimirpetrov ,
tnx

**Comment by rhtyd** [comment]
My suspicion is that 4.14 systemvms (ssvm, cpvm and VRs) don't eject or fail to eject the systemvm.iso file. Can we confirm that @vladimirpetrov ?

**Comment by DaanHoogland** [comment]
ah, if you ar right, that would than pertain to higher upgrade paths as well (4.15 -> future), making this slightly more pressing

**Comment by vladimirpetrov** [comment]
Attaching the log from management server start after upgrade 4.14 -> 4.15.
[management-server.log.2020-11-25.gz](https://github.com/apache/cloudstack/files/5635240/management-server.log.2020-11-25.gz)

**Comment by vladimirpetrov** [comment]
The systemvm.iso file is not detected by the host VM after upgrade:

![image](https://user-images.githubusercontent.com/12384665/100994191-03fcda80-355f-11eb-99b8-365923172940.png)

**Comment by vladimirpetrov** [comment]
> Can you share more logs/info to let others also understand the issue as well.
> 
> Also, I see " _CentOS 7.0 on the hosts._" in the description, but you are saying it's the XS setup - do you have mixed hypervisors or why is there " CentOS 7.0 on the hosts." ?

No, it has only two XenServer hosts, CentOS is just the OS of the hosts and it doesn't seem to be relevant to the problem.

**Comment by vladimirpetrov** [comment]
> Can you share more logs/info to let others also understand the issue as well.
> 
> Also, I see " _CentOS 7.0 on the hosts._" in the description, but you are saying it's the XS setup - do you have mixed hypervisors or why is there " CentOS 7.0 on the hosts." ?

Sorry Andrija, my mistake - I edited the original post to avoid confusion. It's not mixed environment, only 2 Xen hosts.

**Comment by andrijapanicsb** [comment]
This need be fixed before 4.15 can go out.

**Comment by rhtyd** [comment]
I setup 4.14 + XS71 and did an upgrade test and was unable to reproduce this. I suspect in the upgraded env the old ssvm/cpvm were killed and new ssvm/cpvm got started before the 4.15 systemvm.iso/files/scripts were patched on the XS71 hosts which could explain the race conditions around old 4.14 systemvm.iso.

**Comment by rhtyd** [comment]
Lowered from blocker as we don't consistently reproduce this, let's wait for @vladimirpetrov @andrijapanicsb and others if they can confirm it works or doesn't work. Thanks.

**Comment by PaulAngus** [comment]
It’s a blocker even if it happens inconsistently. Until Vladi can confirm that is a non-issue if the operator ensures that all hosts have been updated before recreating system VMs. The potential issue IS still a blocker.

**Comment by andrijapanicsb** [comment]
+1 what Paul said.

Andrija Panic
Cloud Architect
ShapeBlue
________________________________

**Comment by rhtyd** [comment]
I did an upgrade test and discussion with Vladi, the original issue was due to not waiting enough after an upgrade for the management server to copy all the systemvm.iso/files/scripts to the XS host post-upgrade.

In the latest upgrade env created by Vladi, from 4.14->4.15 with XS71, we couldn't reproduce the issue. However we had to restart the systemvms a few times due to a certificate setup issue likely caused due to:
- CPU pressure - XS hosts are run inside VM, some cpu entropy is required to create keystore and certificates - this applies to only to SSVM and CPVM
- Once SSVM and CPVM were up and connected, we were able to deploy a new VM on a new network that came up OK
- We found that VRs failed on XS hosts with already too many VMs - VRs appeared to run slow on these hosts and had errors in running the get_template_version domr script, this didnot happen on XS hosts with less num. of VMs

From experience I've seen VMs runs slow on XS65, 71; they appear to run faster on more recent XS version (it could also be a hvm setting).

@vladimirpetrov can you add your notes/findings and confirmation.

**Comment by rhtyd** [comment]
PR fix https://github.com/apache/cloudstack/pull/4525 pl test and review @vladimirpetrov thanks

**Comment by vladimirpetrov** [comment]
The fix worked - well done, @rhtyd !

**Comment by rhtyd** [comment]
Fixed in #4525, kicking smoketests on health check PR #4131
